bbs_plus/src/setup.rs

```diff
@@ -186,59 +187,46 @@ macro_rules! impl_sig_params {
+            /// an error will be returned.
+            /// Eg. if given messages `m_i`, `m_j`, and `m_k` in the iterator, the commitment converts messages to
             /// scalars and multiplies them by the parameter curve points:
             /// `params.h_0 * blinding + params.h_i * m_i + params.h_j * m_j + params.h_k * m_k`
             /// Computes using multi-scalar multiplication
-            pub fn commit_to_messages(
+            pub fn commit_to_messages<'a, MI>(
                 &self,
-                messages: BTreeMap<usize, &E::ScalarField>,
-                blinding: &E::ScalarField,
-            ) -> Result<E::$group_affine, BBSPlusError> {
-                #[cfg(feature = "parallel")]
-                let (mut bases, mut scalars): (Vec<E::$group_affine>, Vec<E::ScalarField>) = {
-                    // Need to manually check that no message index exceeds the maximum number of messages allowed
-                    // because this function can be called with only uncommitted messages; so size of BTreeMap is
-                    // not representative of the number of messages
-                    for (i, _) in messages.iter() {
-                        if *i >= self.supported_message_count() {
-                            return Err(BBSPlusError::InvalidMessageIdx(*i));
-                        }
-                    }
-                    cfg_into_iter!(messages)
-                        .map(|(i, msg)| (self.h[i].clone(), *msg))
-                        .unzip()
-                };
-
-                #[cfg(not(feature = "parallel"))]
-                let (mut bases, mut scalars): (Vec<E::$group_affine>, Vec<E::ScalarField>) = {
-                    let mut bases = Vec::with_capacity(messages.len());
-                    let mut scalars = Vec::with_capacity(messages.len());
-                    for (i, msg) in messages.into_iter() {
-                        // No message index should be >= max supported messages
-                        if i >= self.supported_message_count() {
-                            return Err(BBSPlusError::InvalidMessageIdx(i));
-                        }
-                        bases.push(self.h[i].clone());
-                        scalars.push(*msg);
-                    }
-                    (bases, scalars)
-                };
-
-                bases.push(self.h_0.clone());
-                scalars.push(*blinding);
+                indexed_messages_sorted_by_index: MI,
+                blinding: &'a E::ScalarField,
+            ) -> Result<E::$group_affine, BBSPlusError>
+            where
+                MI: IntoIterator<Item = (usize, &'a E::ScalarField)>,
+            {
+                let (bases, scalars): (Vec<_>, Vec<_>) = process_results(
+                    pair_valid_pairs_with_slice::<_, _, _, BBSPlusError, _>(
+                        indexed_messages_sorted_by_index,
+                        |(a, _), (b, _)| a < b,
```

Comment:
> It shouldn't matter if the passed message iterator is sorted or not as it already contains the message index.

The current code emulates `BTreeMap` behavior (elements sorted by unique indices). If this check is removed, the caller can provide messages with repeated indices. If it doesn't matter, we can omit validation.
>  Secondly the call to pair_valid_pairs_with_slice will iterate over the passed messages multiple times

How comes? This iterator doesn't even require `Clone`, how it can be iterated multiple times?
All the functions will be composed as per iterator spec, and the final code can be approximated to this equivalent:

```rust
let mut last = None;
let mut res = Vec::new();

for item in indexed_messages_sorted_by_index {
    if last.as_ref().map_or(true, |last| check(last, item)) {
        last.replace(item.clone());
        if let Some(h_i) = h.get(item.0) {
            res.push((h_i, item.1));
            continue;
        }
    }
    
    break;
}
```

Reply:
UPD: I improved pair validation, so the index will be copied only during iteration, not the full item as was implemented previously.